distributed/process.py

```diff
@@ -90,9 +104,42 @@ def _on_exit(self, exitcode):
             self._exit_callback(self)
         self._exit_future.set_result(exitcode)
 
+
+    @classmethod
+    def _immediate_exit_when_closed(cls, parent_alive_pipe):
+        """
+        Immediately exit the process when parent_alive_pipe is closed.
+        """
+        def monitor_parent():
+            try:
+                # The parent_alive_pipe should be held open as long as the
+                # parent is alive and wants us to stay alive. Nothing writes to
+                # it, so the read will block indefinitely.
+                parent_alive_pipe.recv()
+            except EOFError:
+                # Parent process went away unexpectedly. Exit immediately. Could
+                # consider other exiting approches here. My initial preference
+                # is to unconditionally and immediately exit. If we're in this
+                # state it is possible that a "clean" process exit won't work
+                # anyway - if, for example, the system is getting bogged down
+                # due to the running out of memory, exiting sooner rather than
+                # later might be needed to restore normal system function.
+                # If this is in appropriate for your use case, please file a
+                # bug.
+                os._exit(-1)
```

Comment:
Hm, this is a tricky one. Partially, I made this choice of exiting because it is the easy choice: it works. (A quick attempt with the other methods don't work as I'd expect, so there would be more work to do to make this work.)

It is correct that atexit handlers don't run; `finalizers` also don't run. On the one hand this is bad (in the case of temporary files it's another potential resource leak). On the other hand it gets us to our intended final state more quickly (not hogging resources). It's also the case that this is currently used in an extreme situation: the parent process has gone unexpectedly. Maybe the child processes should follow suit?

I suppose you're talking about the pid files, yes? As far as I can tell, those should be unaffected, because the process which is `_exit`ing is a thing managed by an `AsyncProcess`, not the main thread. The main thread in my use case has been `Killed` - so would leak its pidfile in any case, if you had one.